aquadoggo/src/graphql/schema_builders/document.rs

```diff
@@ -0,0 +1,474 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use async_graphql::dynamic::{Field, FieldFuture, InputValue, Object, TypeRef};
+use dynamic_graphql::{Error, FieldValue, ScalarValue};
+use log::debug;
+use p2panda_rs::storage_provider::traits::DocumentStore;
+use p2panda_rs::{document::traits::AsDocument, schema::Schema};
+
+use crate::db::SqlStore;
+use crate::graphql::scalars::{DocumentIdScalar, DocumentViewIdScalar};
+use crate::graphql::types::DocumentMeta;
+use crate::graphql::utils::{downcast_id_params, fields_name, get_document_from_params};
+
+const QUERY_ALL_PREFIX: &str = "all_";
+const FIELDS_FIELD: &str = "fields";
+const META_FIELD: &str = "meta";
+const DOCUMENT_META_SCHEMA_ID: &str = "DocumentMeta";
+const DOCUMENT_ID_ARG: &str = "id";
+const DOCUMENT_VIEW_ID_ARG: &str = "viewId";
+const DOCUMENT_ID_SCALAR: &str = "DocumentId";
+const DOCUMENT_VIEW_ID_SCALAR: &str = "DocumentViewId";
+
+/// Build a graphql object type for a p2panda schema.
+///
+/// Contains resolvers for both `fields` and `meta`. The former simply passes up the query
+/// arguments to it's children query fields. The latter retrieves the document being queried and
+/// already constructs and returns the `DocumentMeta` object.
+pub fn build_document_schema(schema: &Schema) -> Object {
+    let document_fields_name = fields_name(schema.id());
+    Object::new(schema.id().to_string())
+        // The `fields` of a document, passes up the query arguments to it's children.
```

Comment:
I stumbled upon this a little, maybe it should say "The `fields` field of a GraphQL document"? In general the GraphQL terminology mixes quite up with p2panda, but I guess there is not much we can do 😆

Reply:
Lolz, I nearly wrote "`fields` field", will change it to that.